File: step-01-basics/basic_recursion/sol.cpp
// problem :
// link    :
// pattern :
// brute   : O( ) time, O( ) space
// better  : O( ) time, O( ) space
// optimal : O( ) time, O( ) space

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <queue>
#include <stack>
#include <climits>
#include <cmath>
using namespace std;

void nameNtimes(int n, string name)
{
    if (n == 0)
    {
        return;
    }
    cout << name << endl;
    nameNtimes(n - 1, name);
}

void print1ToN(int n)
{
    if (n == 0)
    {
        return;
    }
    print1ToN(n - 1);
    cout << n << endl;
}

void printNto1(int n)
{
    if (n == 0)
    {
        return;
    }
    cout << n << endl;
    printNto1(n - 1);
}

int nNumberNaturalSum(int n)
{
    if (n == 1)
    {
        return n;
    }
    return n + nNumberNaturalSum(n - 1);
}

int factorial(int n)
{
    if (n == 1 || n == 0)
    {
        return 1;
    }
    return n * factorial(n - 1);
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    nameNtimes(3, "Ritvik");
    print1ToN(5);
    printNto1(5);
    cout << nNumberNaturalSum(100) << endl;
    cout << factorial(5) << endl;

    return 0;
}

File: step-01-basics/basic_recursion/sol.py
"""
problem :
link    :
pattern :
brute   : O( ) time, O( ) space
optimal : O( ) time, O( ) space
"""
from typing import List, Optional
from collections import defaultdict, deque, Counter
from heapq import heappush, heappop, heapify
from bisect import bisect_left, bisect_right

def nameNtimes(name, n):
    if(n==0):
        return
    print(name)
    nameNtimes(name,n-1)

def oneToN(n):
    if(n==0):
        return
    oneToN(n-1)
    print(n)

def nToOne(n):
    if(n==0):
        return
    print(n)
    nToOne(n-1)

def nNaturalSum(n):
    if(n==1):
        return n
    return n + nNaturalSum(n-1)

def factorial(n):
    if(n==1 or n==0):
        return 1
    return n*factorial(n-1)

def solve():
    nameNtimes("Ritvik",3)
    oneToN(5)
    nToOne(5)
    print(nNaturalSum(100))
    print(factorial(5))

if __name__ == "__main__":
    solve()
